components/sendgrid/actions/list-global-supressions/list-global-supressions.js

```diff
@@ -0,0 +1,77 @@
+const validate = require("validate.js");
+const common = require("../common");
+
+module.exports = {
+  ...common,
+  key: "sendgrid-list-global-supressions",
+  name: "List Global Supressions",
+  description:
+    "Allows you to get a list of all email address that are globally suppressed.",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    ...common.props,
+    startTime: {
+      type: "integer",
+      label: "Start Time",
+      description:
+        "Refers start of the time range in unix timestamp when an unsubscribe email was created (inclusive).",
+      optional: true,
+    },
+    endTime: {
+      type: "integer",
+      label: "End Time",
+      description:
+        "Refers end of the time range in unix timestamp when an unsubscribe email was created (inclusive).",
+      optional: true,
+    },
+    numberOfSupressions: {
+      type: "integer",
+      label: "Max # of Global Supressions to Return",
+      description: "Indicates the max number of global supressions to return.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
+  async run() {
+    const constraints = {
+      numberOfSupressions: {
+        type: "integer",
+      },
+    };
+    if (this.startTime) {
+      constraints.startTime = this.getIntegerGtZeroConstraint();
+    }
+    if (this.endTime) {
+      constraints.endTime = {
+        numericality: {
+          onlyInteger: true,
+          greaterThan: this.startTime > 0 ?
+            this.startTime :
+            0,
+          message: "must be positive integer, non zero, greater than `startTime`.",
+        },
+      };
+    }
+    if (this.numberOfSupressions) {
+      constraints.numberOfSupressions = this.getIntegerGtZeroConstraint();
+    }
+    const validationResult = validate(
+      {
+        startTime: this.startTime,
+        endTime: this.endTime,
+        numberOfSupressions: this.numberOfSupressions,
+      },
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    const listSupressionsEndpoint = "/v3/suppression/unsubscribes";
+    return this.sendgrid.listItems(
+      listSupressionsEndpoint,
+      this.startTime,
+      this.endTime,
+      this.numberOfSupressions,
+    );
```

Comment:
To encapsulate knowledge of the API in the [app file](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#good-api-usage), I'd suggest moving the `listBlocksEndpoint` (`/v3/suppression/unsubscribes`) to the app file so it can be reused. For example:

- In `sendgrid.app.js`
```js
    async listUnsubscribes(startTime, endTime, maxItems) {
      const listItemsEndpoint = "/v3/suppression/unsubscribes";
      return this.listItems(listItemsEndpoint, startTime, endTime, maxItems);
    },
```

- In `list-global-supresions.js`
```js
    this.checkValidationResults(validationResult);
    return this.sendgrid.listUnsubscribes(
      this.startTime,
      this.endTime,
      this.numberOfSuppressions,
    );
```

Reply:
Ok, I have implemented this as well, however. I preserved the method name `listGlobalSupressions` to be consistent with the API documentation, where the API reference entry is named "Retrieve all global suppressions " but the related API endpoint is related to unsubscribes:

> Retrieve all global suppressions
> GET /v3/suppression/unsubscribes

See [Retrieve all global suppressions](https://docs.sendgrid.com/api-reference/suppressions-global-suppressions/retrieve-all-global-suppressions) at the Sendgrid's API Referce.

But please let me know if keeping listUnsubscribesis better for user readiability or any other reason.